tests/test_clients/mod.rs

```diff
@@ -0,0 +1,305 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+pub mod raw_request;
+pub mod stress;
+
+pub use raw_request::RawRequestClient;
+
+pub use parsec_client::core::request_client::RequestClient;
+pub use parsec_client::error;
+
+use log::error;
+use parsec_client::auth::AuthenticationData;
+use parsec_client::core::basic_client::BasicClient;
+use parsec_client::error::Result;
+use parsec_interface::operations::list_providers::ProviderInfo;
+use parsec_interface::operations::psa_algorithm::{Algorithm, AsymmetricSignature, Hash};
+use parsec_interface::operations::psa_key_attributes::*;
+use parsec_interface::operations::psa_key_attributes::{KeyAttributes, KeyPolicy, UsageFlags};
+use parsec_interface::requests::{Opcode, ProviderID};
+use std::collections::HashSet;
+
+/// Client structure automatically choosing a provider and high-level operation functions.
+#[derive(Debug)]
+pub struct TestClient {
+    basic_client: BasicClient,
+    created_keys: Option<HashSet<(String, String, ProviderID)>>,
+}
+
+impl TestClient {
+    /// Creates a TestClient instance.
+    ///
+    /// The implicit provider chosen for servicing cryptographic operations is decided through
+    /// a call to `list_providers`, followed by choosing the first non-Core provider.
+    pub fn new() -> TestClient {
+        let mut client = TestClient {
+            basic_client: BasicClient::new(
+                AuthenticationData::AppIdentity(String::from("root")),
+                ProviderID::Core,
+            ),
```

Comment:
This constructor feels a bit like the chicken and the egg: you need an implicit provider for non-core operations but you can only do that with `list_providers` and you need the basic client instance for that.
That is a client change I did not notice before but what do you think if we either take the following decisions:
1. put the first-crypto provider discovery inside the `BasicClient` constructor. Not so basic now...
2. remove the provider argument from the constructor and always but `Core`, with doc comments saying that `list_providers` is the first thing expected.
3. remove the provider argument and put an `Option<ProviderID>` in the `BasicCLient` structure, initialised at `None` by the constructor. We also create a new client error which is throwm when no `implicit_provider` has been selected.

I personally like 3 best.

Reply:
Maybe we should continue this in the appropriate GitHub issue

Reply:
We can try and settle it here, as it's not entirely clear if it's something lacking in the `BasicClient` or here.

3 makes sense, but I'm guessing you wouldn't want to change the `implicit_provider` back to `None`. I think that matching could be done efficiently, so I can do that as another PR in the client.

Reply:
I guess if `implicit_provider` is private, you would only change it to `Some(ProviderID::...)` through methods. Also checking that it's not changing it to `Core`.
One more suggestion to make it even more clear in the client: renaming `implicit_provider` to `implicit_crypto_provider` which what it really is!

Those things give me the feeling that there would be even more reasons to separate our structures (opcodes, provide trait, provider ID, response status) between Crypto (PSA) and Core ones.

Reply:
> One more suggestion to make it even more clear in the client: renaming implicit_provider to implicit_crypto_provider which what it really is!

Reply:
I think I'll keep it as `implicit_provider` - there will be operations that are not explicitly crypto but which need a provider defined (the smart defaulting stuff) - and while they're not available now, it doesn't hurt to have them ready. As you were saying, I'd rather give the user an option to set it to whatever they wish, regardless of whether it is allowed for currently supported operations. It feels like more of a problem with our `ProviderID` definition than with this field

Reply:
> there will be operations that are not explicitly crypto

Those operations will still be aimed at a a crypto provider though, right? They can still change it with the `set_crypto_provider_id` methods I think.
My suggestion was mainly about removing the provider ID argument for the `new` method.

Reply:
I would be happy with no renaming at all but just the `implicit_provider` being an `Option<ProviderID>`, the `new` method setting it at `None` and crypto methods returning a new error if it is `None`.
No need to make more checks in the `set_provider` method for now.